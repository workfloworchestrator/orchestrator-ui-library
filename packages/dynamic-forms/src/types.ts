import React, { FormEventHandler } from 'react';
import { FieldValues, useForm } from 'react-hook-form';

import { z } from 'zod';

export type TDynamicFormFieldOptionsType = 'anyOf' | 'allOf';

export enum DfFieldFormats {
    DATE = 'date',
    HIDDEN = 'hidden',
    DATETIME = 'date-time',
    SKIP = 'skip',
    TIMESTAMP = 'timestamp',

    DEFAULT = 'text',
    DROPDOWN = 'dropdown',
    BOOLFIELD = 'boolfield',
    SELECT = 'select',
    MULTISELECT = 'multiselect',
    CHECKBOX = 'checkbox',
    RADIO = 'radio',
    SWITCH = 'switch',
    DATEPICKER = 'datepicker',
    NUMBER = 'number',
    LIST = 'list',
}

export enum DfFieldTypes {
    OPTGROUP = 'optGroup',
    SKIP = 'skip',
    LONG = 'long',
    HIDDEN = 'hidden',
    LABEL = 'label',
    SUMMARY = 'summary',
    ACCEPT = 'accept',
    NULL = 'null',

    DATE = 'date',
    DATETIME = 'date-time',
    OBJECT = 'object',
    STRING = 'string',
    ARRAY = 'array',
    BOOLEAN = 'boolean',
    NUMBER = 'number',
    TIMESTAMP = 'timestamp',
    LIST = 'list',
    DEFINED_LIST = 'defined-list',
    DEFINED_MULTI_LIST = 'defined-multi-list',
    TWO_OPTION_LIST = 'two-options-list',
}

export enum DFFieldReturnType {
    STRING = 'string',
    NUMBER = 'number',
    DATETIME = 'date-time',
}
export interface IValidationErrorDetails {
    detail: string;
    source: IDynamicFormApiValidationError[];
    mapped: {
        [fieldId: string]: {
            value: string;
            msg: string;
        };
    };
}

export interface IDynamicFormsContextProps {
    isLoading: boolean;
    isSending: boolean;
    isFullFilled: boolean;
    theReactHookForm: ReturnType<typeof useForm>;
    errorDetails?: IValidationErrorDetails;
    formData?: IDynamicForm;
    debugMode?: boolean;
    title?: string;
    sendLabel?: string;
    onCancel?: () => void;
    submitForm: FormEventHandler<HTMLFormElement>;
    resetForm: (e: MouseEvent) => void;
    successNotice?: React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;
}

export type DynamicFormsMetaData = {
    [key: string | number]: DfFieldValue;
};

export type DfFormProvider = ({
    formKey,
    requestBody,
}: {
    formKey: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    requestBody: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfLabelProvider = ({
    formKey,
    id,
}: {
    formKey: string;
    id?: string | null;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfDataProvider = () => Promise<IDynamicFormsLabels>;

export type CustomValidationRuleFn = (
    fieldConfig: IDynamicFormField,
    theReactHookForm?: ReturnType<typeof useForm>,
) => Zod.ZodTypeAny | undefined;

export interface IDynamicFormsContextConfig {
    // use custom method to provide data for the form. This overwrites data fetched from labels endpoint
    dataProvider: DfDataProvider;

    // use custom method for providing labels and data
    labelProvider: DfLabelProvider;

    // use a custom method for providing the form definition
    formProvider: DfFormProvider;

    // Extend field definitions
    fieldDetailProvider?: IFieldDefinitionProvider;

    // be able to refresh the provided data
    dataProviderCacheKey?: number;

    // whenever a fieldvalue changes, do something
    onFieldChangeHandler?: onFieldChangeHandlerFn;

    // provide custom validation rules for fields
    customValidationRules?: CustomValidationRuleFn;

    // This is a temp solution for 2 different Backend implementations of pydantic forms
    // we will remove this once the backends both work the same
    // Portal backend has a hack that is a leftover from previous implementation
    // So for only that backend, we should not send the data as a array
    tmp_pydanticFormsOriginalImplementation?: boolean;
}

export interface IDynamicFormsContextInitialProps {
    formKey: string;
    formIdKey?: string;
    title?: string;
    sendLabel?: string;
    metaData?: DynamicFormsMetaData;
    successNotice?: React.ReactNode;
    onSuccess?: (fieldValues: FieldValues) => void;
    onCancel?: () => void;
    children: (props: IDynamicFormsContextProps) => React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;

    config: IDynamicFormsContextConfig;
}

export interface IFieldDefinitionProvider {
    [fieldId: string]: Partial<IDynamicFormField>;
}

export type onFieldChangeHandlerFn = (
    field: {
        name?: string;
        type?: string;
        value: DfFieldValue;
    },
    theReactHookForm: ReturnType<typeof useForm>,
) => void;

export interface IDynamicFormApiResponseDefEnum {}

export interface TJsonSchemaRef {
    $ref: string;
}

export interface TDynamicFormFieldAnyOfDef {
    items?: TJsonSchemaRef;
    type: 'null' | 'array';
}

export type ListFieldType =
    | DfFieldTypes.STRING
    | DfFieldTypes.DATE
    | DfFieldTypes.NUMBER
    | DfFieldTypes.NULL;

export interface ITDynamicFormFieldAnyOfResolvedItems {
    enum: string[];
    title: string;
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface TDynamicFormFieldAnyOfResolved {
    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];
    format?: 'date' | 'date-time';
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseProperty {
    type: DfFieldTypes;
    anyOf?: TDynamicFormFieldAnyOfDef[];
    default?: string | null;
    title: string;
}

export interface IDynamicFormApiResponsePropertyResolved {
    type: DfFieldTypes;

    anyOf?: TDynamicFormFieldAnyOfResolved[];
    oneOf?: TDynamicFormFieldAnyOfResolved[];
    allOf?: TDynamicFormFieldAnyOfResolved[];

    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];

    default?: string | null;
    title: string;
    format: DfFieldFormats;

    uniforms?: {
        disabled: boolean;
    };

    // validation props
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseBase {
    title: string;
    description: string;
    additionalProperties: boolean;
    type: 'object';
    required?: string[];
    $defs?: {
        [definitionId: string]: {
            enum: string[];
            title: string;
            type: DfFieldTypes;
        };
    };
}

export interface IDynamicFormApiResponse extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponseProperty;
    };
}

export interface IDynamicFormApiErrorResponse {
    detail?: string;
    status: number;
    form: IDynamicFormApiResponse;
    success?: boolean;
    validation_errors: IDynamicFormApiValidationError[];
}

export interface IDynamicFormApiValidationError {
    input: string;
    loc: string[];
    msg: string;
    type: string;
    url: string; //"https://errors.pydantic.dev/2.4/v/extra_forbidden"
}

export interface IDynamicFormApiRefResolved
    extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponsePropertyResolved;
    };
}

export interface IDynamicFormsLabels {
    [key: string]: string[] | number[] | string | number | null;
}

export interface IDynamicFormsLabelResponse {
    labels: IDynamicFormsLabels;
}

export interface IDynamicFormFieldValidation {
    maxLength?: number;
    minLength?: number;
    pattern?: string;
    isNullable?: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DfFieldValue = any;

export interface IDynamicFormFieldAttributes {
    disabled?: boolean;
}

export interface IDynamicFormField {
    id: string;
    title: string;
    description: string;
    type: DfFieldTypes;
    format: DfFieldFormats;
    options: string[];
    disabledOptions?: string[];
    default?: DfFieldValue;
    required: boolean;
    isEnumField: boolean;
    schemaField: IDynamicFormApiResponsePropertyResolved;
    validation: IDynamicFormFieldValidation;
    attributes: IDynamicFormFieldAttributes;
    validator?: FormZodValidationFn;
    FormElement?: FormElementComponent;
    matchedFieldResult?: DfFieldConfig;
}

export interface IDynamicFormFieldSection {
    id: string;
    title: string;
    fields: string[];
}

export enum IDynamicFormState {
    NEW = 'new',
    IN_PROGRESS = 'in-progress',
    FINISHED = 'finished',
}

export interface IDynamicForm {
    title: string;
    description: string;
    state: IDynamicFormState;
    fields: IDynamicFormField[];
    sections: IDynamicFormFieldSection[];
}

export type FormZodValidationFn = (field: IDynamicFormField) => z.ZodTypeAny;
export type FormElementComponent = (props: IDFInputFieldProps) => JSX.Element;

export interface FormComponent {
    Element: FormElementComponent;
    validator?: FormZodValidationFn;
}

export interface DfFieldConfig {
    id: string;
    Component: FormComponent;
    matcher?: (field: IDynamicFormField) => boolean;
    preventColRender?: boolean;
}

export type DfFieldsConfig = DfFieldConfig[];

export interface IDFInputFieldProps {
    field: IDynamicFormField;
}

export interface IDFZodValidationPresets {
    [type: string]: FormZodValidationFn;
}
