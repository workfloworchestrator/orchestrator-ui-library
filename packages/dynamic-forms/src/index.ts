import DynamicForm from "./components/form/ContextWrap"

export default DynamicForm
