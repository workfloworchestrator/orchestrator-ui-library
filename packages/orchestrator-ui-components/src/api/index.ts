/*
 * Copyright 2019-2023 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { AxiosInstance } from 'axios';

import {
    ImsPort,
    IpBlock,
    IpPrefix,
    NodeSubscription,
} from '../components/WfoForms/formFields/surf/types';
import { ProductDefinition } from '../types';
import { getAxiosInstance } from './axios';

const FORMS_ENDPOINT = 'surf/forms/';
const PROCESS_ENDPOINT = 'processes/';
const PRODUCTS_ENDPOINT = 'products/';

export class BaseApiClient {
    private _axiosInstance: AxiosInstance;

    constructor(apiPath: string) {
        this._axiosInstance = getAxiosInstance(apiPath);
    }

    axiosFetch = <R = object>(
        path: string,
        options = {},
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        headers = {},
        showErrorDialog = true,
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = true,
    ): Promise<R> => {
        // preset the config with the relative URL and a GET type.
        // presets can be overridden with `options`.
        return this._axiosInstance({ url: path, method: 'GET', ...options })
            .then((res) => res.data)
            .catch((err) => {
                if (showErrorDialog) {
                    setTimeout(() => {
                        throw err;
                    }, 250);
                }
                throw err;
            });
    };

    catchErrorStatus = <T>(
        promise: Promise<unknown>,
        status: number,
        callback: (json: T) => void,
    ) => {
        return promise.catch((err) => {
            if (err.response && err.response.status === status) {
                callback(err.response.data);
            } else {
                throw err;
            }
        });
    };

    fetchJson = <R = object>(
        path: string,
        options = {},
        headers = {},
        showErrorDialog = true,
        result = true,
    ): Promise<R> => {
        return this.axiosFetch(path, options, headers, showErrorDialog, result);
    };

    fetchJsonWithCustomErrorHandling = <R = object>(
        path: string,
    ): Promise<R> => {
        return this.fetchJson(path, {}, {}, false, true);
    };

    postPutJson = <R = object>(
        path: string,
        processInput: object,
        method: string,
        showErrorDialog = true,
        result = true,
    ): Promise<R> => {
        return this.axiosFetch(
            path,
            { method: method, data: processInput },
            {},
            showErrorDialog,
            result,
        );
    };
}

abstract class ApiClientInterface extends BaseApiClient {
    abstract cimStartForm: (
        formKey: string,
        userInputs: object[],
    ) => Promise<object>;
}

export class ApiClient extends ApiClientInterface {
    startProcess = (
        workflowName: string,
        processInput: object,
    ): Promise<unknown> => {
        return this.postPutJson<unknown>(
            `${PROCESS_ENDPOINT}${workflowName}`,
            processInput,
            'post',
            false,
            true,
        );
    };

    resumeProcess = (processId: string, userInput: object[]) => {
        return this.postPutJson(
            `processes/${processId}/resume`,
            userInput,
            'put',
            false,
            false,
        );
    };

    products = (): Promise<ProductDefinition[]> => {
        return this.fetchJson<ProductDefinition[]>(PRODUCTS_ENDPOINT);
    };
    productById = (productId: string): Promise<ProductDefinition> => {
        return this.fetchJson(`${PRODUCTS_ENDPOINT}${productId}`);
    };
    cimStartForm = (
        formKey: string,
        userInputs: object[],
    ): Promise<{ id: string }> => {
        return this.postPutJson(
            `${FORMS_ENDPOINT}${formKey}`,
            userInputs,
            'post',
            false,
            true,
        );
    };
    prefix_filters = (): Promise<IpPrefix[]> => {
        return this.fetchJson('surf/ipam/prefix_filters');
    };
    ip_blocks = (parentPrefix: number): Promise<IpBlock[]> => {
        return this.fetchJson('surf/ipam/ip_blocks/' + parentPrefix);
    };
    free_subnets = (
        subnet: string,
        netmask: number,
        prefixlen: number,
    ): Promise<string[]> => {
        return this.fetchJson(
            'surf/ipam/free_subnets/' +
                subnet +
                '/' +
                netmask +
                '/' +
                prefixlen,
        );
    };
    getFreePortsByNodeSubscriptionIdAndSpeed = (
        nodeSubscriptionId: string,
        interfaceSpeed: number,
        mode: string,
    ): Promise<ImsPort[]> => {
        return this.fetchJson(
            `surf/ims/free_ports/${nodeSubscriptionId}/${interfaceSpeed}/${mode}`,
        );
    };

    // legacy : for imsPort selector
    subscriptions = (
        tagList: string[] = [],
        statusList: string[] = [],
        productList: string[] = [],
    ): Promise<NodeSubscription[]> => {
        const filters = [];

        if (tagList.length)
            filters.push(`tags,${encodeURIComponent(tagList.join('-'))}`);
        if (statusList.length)
            filters.push(
                `statuses,${encodeURIComponent(statusList.join('-'))}`,
            );
        if (productList.length)
            filters.push(
                `products,${encodeURIComponent(productList.join('-'))}`,
            );

        const params = new URLSearchParams();
        if (filters.length) params.set('filter', filters.join(','));

        return this.fetchJson(
            `subscriptions/${filters.length ? '?' : ''}${params.toString()}`,
        );
    };

    nodeSubscriptions = (
        statusList: string[] = [],
    ): Promise<NodeSubscription[]> => {
        return this.subscriptions(['Node'], statusList);
    };
}

export function getApiClient(apiEndPoint: string) {
    return new ApiClient(apiEndPoint);
}
