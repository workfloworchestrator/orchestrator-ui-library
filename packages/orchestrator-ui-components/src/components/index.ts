export * from './WfoBadges';
export * from './WfoPageTemplate/WfoBreadcrumbs';
export * from './WfoFilterTabs';
export * from './WfoKeyValueTable';
export * from './WfoPageTemplate';
export * from './WfoSearchBar';
export * from './WfoSettingsModal';
export * from './WfoSettingsPage';
export * from './WfoStartPage';
export * from './WfoSubscription';
export * from './WfoTable';
export * from './WfoTimeline';
export * from './WfoLoading';
export * from './WfoToastsList';
export * from './WfoForms';
export * from './WfoStartTaskButtonComboBox';
