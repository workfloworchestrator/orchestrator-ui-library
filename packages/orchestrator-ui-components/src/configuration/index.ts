export * from './constants';
export * from './policy-resources';
