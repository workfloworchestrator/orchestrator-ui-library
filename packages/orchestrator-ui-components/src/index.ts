export * from './lib/components';

export * from './lib/icons';

export * from './lib/theme';

export * from './lib/utils';

export * from './lib/contexts';

export * from './lib/types';

export * from './lib/hooks';
