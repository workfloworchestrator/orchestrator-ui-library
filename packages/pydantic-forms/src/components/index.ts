export * from './forms';
export * from './formFields';
