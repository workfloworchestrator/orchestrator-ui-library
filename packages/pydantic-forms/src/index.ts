export * from './components/index';
export * from './contexts/index';
export * from './messages/index';
export * from './types/index';
