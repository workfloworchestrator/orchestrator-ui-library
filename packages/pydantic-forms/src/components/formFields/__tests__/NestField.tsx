/*
 * Copyright 2019-2023 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import createContext from "lib/uniforms-surfnet/__tests__/_createContext";
import mount from "lib/uniforms-surfnet/__tests__/_mount";
import { ListField, NestField } from "lib/uniforms-surfnet/src";
import React from "react";
import { AutoField } from "uniforms-unstyled";

test("<NestField> - renders an <AutoField> for each field", () => {
    const element = <NestField name="x" id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext(
            {
                x: { type: Object },
                "x.a": { type: String },
                "x.b": { type: Number },
            },
            undefined,
            "nestfield-autofield"
        )
    );

    expect(wrapper.find(AutoField)).toHaveLength(2);
    expect(wrapper.find(AutoField).at(0).prop("name")).toBe("a");
    expect(wrapper.find(AutoField).at(1).prop("name")).toBe("b");
    expect(wrapper.render()).toMatchSnapshot();
});

test("<NestField> - renders custom content if given", () => {
    const element = (
        <NestField name="x">
            <article data-test="content" />
        </NestField>
    );
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.a": { type: String },
            "x.b": { type: Number },
        })
    );

    expect(wrapper.find(AutoField)).toHaveLength(0);
    expect(wrapper.find("article")).toHaveLength(1);
    expect(wrapper.find("article").prop("data-test")).toBe("content");
});

test("<NestField> - renders a label", () => {
    const element = <NestField name="x" label="y" />;
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.a": { type: String },
            "x.b": { type: Number },
        })
    );

    expect(wrapper.find("label")).toHaveLength(2);
    expect(wrapper.find("span.euiTitle").at(0).text()).toBe("y");
});

test("<NestField> - renders a wrapper with unknown props", () => {
    const element = <NestField name="x" data-x="x" data-y="y" data-z="z" />;
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.a": { type: String },
            "x.b": { type: Number },
        })
    );

    expect(wrapper.find(".nest-field").at(0).prop("data-x")).toBe("x");
    expect(wrapper.find(".nest-field").at(0).prop("data-y")).toBe("y");
    expect(wrapper.find(".nest-field").at(0).prop("data-z")).toBe("z");
});

test("<NestField> - renders correctly in list", () => {
    const element = <ListField name="x" id="snapshot-test" />;
    const wrapper = mount(
        element,
        createContext(
            {
                x: { type: Array },
                "x.$": { type: Object },
                "x.$.a": { type: String },
                "x.$.b": { type: Number },
            },
            undefined,
            "nestfield-list"
        )
    );

    expect(wrapper.render()).toMatchSnapshot();
});
