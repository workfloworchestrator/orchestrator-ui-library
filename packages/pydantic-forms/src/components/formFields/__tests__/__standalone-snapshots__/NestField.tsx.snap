// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NestField> - renders an <AutoField> for each field 1`] = `
LoadedCheerio {
  "0": <div
    class="euiDescribedFormGroup  nest-field"
    id="snapshot-test"
    required=""
    role="group"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem"
      >
        <span
          class="euiTitle euiTitle--xsmall euiDescribedFormGroup__title"
        />
      </div>
      <div
        class="euiFlexItem euiDescribedFormGroup__fields euiDescribedFormGroup__fieldPadding--xsmall"
      >
        <section
          required=""
        >
          <div
            class="euiFormRow euiFormRow--fullWidth"
            id="nestfield-autofield-0001-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                class="euiFormLabel euiFormRow__label"
                for="nestfield-autofield-0001"
                id="nestfield-autofield-0001-label"
              />
               
              <div
                class="euiText euiText--medium"
              />
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--fullWidth"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldText euiFieldText--fullWidth"
                    id="nestfield-autofield-0001"
                    name="x.a"
                    placeholder=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </section>
        <div
          required=""
        >
          <div
            class="euiFormRow euiFormRow--fullWidth"
            id="nestfield-autofield-0002-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                class="euiFormLabel euiFormRow__label"
                for="nestfield-autofield-0002"
                id="nestfield-autofield-0002-label"
              />
               
              <div
                class="euiText euiText--medium"
              />
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber"
                    id="nestfield-autofield-0002"
                    name="x.b"
                    placeholder=""
                    step="1"
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<NestField> - renders correctly in list 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive  nest-field"
            id="nestfield-list-0004"
          >
            <div
              class="euiFlexItem"
            >
              <section
                required=""
              >
                <div
                  class="euiFormRow euiFormRow--fullWidth"
                  id="nestfield-list-0005-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="nestfield-list-0005"
                      id="nestfield-list-0005-label"
                    />
                     
                    <div
                      class="euiText euiText--medium"
                    />
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout euiFormControlLayout--fullWidth"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          class="euiFieldText euiFieldText--fullWidth"
                          id="nestfield-list-0005"
                          name="x.0.a"
                          placeholder=""
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </section>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                required=""
              >
                <div
                  class="euiFormRow euiFormRow--fullWidth"
                  id="nestfield-list-0006-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="nestfield-list-0006"
                      id="nestfield-list-0006-label"
                    />
                     
                    <div
                      class="euiText euiText--medium"
                    />
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          class="euiFieldNumber"
                          id="nestfield-list-0006"
                          name="x.0.b"
                          placeholder=""
                          step="1"
                          type="number"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="del-item"
            id="nestfield-list-0007.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <div
          class="add-item"
          id="nestfield-list-0009"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label />
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
